File: java/compcoding/src/main/java/com/kartik/compcoding/ValidParantheses.java
package com.kartik.compcoding;

import java.util.Arrays;
import java.util.Stack;

class ValidParantheses {

    private static boolean isOpen(Character c) {
        return Arrays.asList(new Character[]{'{', '(', '['}).contains(c);
    }

    private static boolean isMatching(Character open, Character close) {
        switch (open) {
            case '{':
                return close == '}';
            case '(':
                return close == ')';
            case '[':
                return close == ']';
            default:
                return false;
        }
    }

    static boolean isValid(String s) {
        Stack<Character> characterStack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            Character currChar = s.charAt(i);
            if (isOpen(currChar)) {
                characterStack.push(currChar);
            } else if (!characterStack.isEmpty() && isMatching(characterStack.peek(), currChar)) {
                characterStack.pop();
            } else {
                return false;
            }
        }
        return characterStack.isEmpty();
    }
}

File: javascript/comp-coding/valid_parantheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
    let stack = [];

    stack[0] = s[0];

    for ( let i = 1; i < s.length; i++) {
        const top = stack[stack.length - 1];

        const currentC = s[i];

        stack = updateStack(stack, top, currentC);
    }
    return stack.length == 0;
};


function updateStack(s, t, cC) {
    console.log(s, t, cC);

    if (cC == '(') {
        if (t == ')') {
            s.pop();
        } else {
            s.push(cC);
        }
    } else if (cC = '[') {
        if (t == ']') {
            s.pop();
        } else {
            s.push(cC);
        }
    } else if (cC = '{') {
        if (t == '}') {
            s.pop();
        } else {
            s.push(cC);
        }
    } else if (cC = ']') {
        if (t == '[') {
            s.pop();
        } else {
            s.push(cC);
        }
    } else if (cC = ')') {
        if (t == '(') {
            s.pop();
        } else {
            s.push(cC);
        }
    } else if (cC = '}') {
        if (t == '{') {
            s.pop();
        } else {
            s.push(cC);
        }
    } else {
        s.push(cC);
    }

    console.log(s);

    return s;
}


console.log(isValid("()"));